lib/code-caller-docker.js

```diff
@@ -0,0 +1,685 @@
+            }
+
+        })(callback);
+    }
+
+    /**
+     * 
+     * @param {'restart' | 'question' | 'course-element' | 'core-element'} type 
+     * @param {string} directory 
+     * @param {string} file 
+     * @param {string} fcn 
+     * @param {string[]} args 
+     * @param {(err: Error | null | undefined, data?: any, output?: string) => void} callback 
+     */
+    call(type, directory, file, fcn, args, callback) {
+        debug(`enter call(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+        this.ensureChild().then(() => {
+            if (!this._checkState([WAITING])) return callback(new Error('invalid DockerCaller state'));
+            if (!this._checkReadyForCall()) return callback(new Error('not ready for call'));
+            // TODO: state checking might be off. for the 'course' allocation method, we only start
+            // containers when we know about the course. however, for the general case in prod, we can
+            // start containers immediately on boot by "warming up". Need to make sure we actually do that,
+            // and that it's handled correctly.
+
+            let cwd = null;
+            let paths = ['/python'];
+            if (type === 'question') {
+                cwd = path.posix.join('/course', 'questions', directory);
+                paths.push('/course/serverFilesCourse');
+            } else if (type === 'course-element') {
+                cwd = path.join('/course', 'elements', directory);
+                paths.push('/course/serverFilesCourse');
+            } else if (type === 'core-element') {
+                cwd = path.posix.join('/elements', directory);
+            } else if (type === 'restart') {
+                // Doesn't need a working directory
+            } else {
+                callback(new Error(`Unknown function call type: ${type}`));
+                return;
+            }
+
+            const callData = { file, fcn, args, cwd, paths };
+            const callDataString = JSON.stringify(callData);
+
+            this.output = '';
+            this.lastCallData = callData;
+            this.callback = callback;
+            // TODO: how long should timeout actually be?
+            // TODO: I think the best option here is to make this timeout longer (1.5x?)
+            // than the timeout in the worker. Starting python processes is cheap,
+            // but starting a new Docker container is relatively expensive. By allowing
+            // the container worker to timeout on bad python code, we'll receive that as
+            // a normal error during call(). The timeout at this level of the code should
+            // only kick in if a container goes completely unresponsive, at which point we
+            // have no choice but to kill it and start over.
+            // We can probably make the Docker worker take the timeout as an option so that
+            // we don't have to coordinate hardcoded timeouts in two places.
```

Comment:
Agreed on passing timeout.

Reply:
This is now provided to the container via a `QUESTION_TIMEOUT_MILLISECONDS` environment variable.